atlasdb-config/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java

```diff
@@ -64,30 +66,35 @@ public FailoverFeignTarget(Collection<String> servers, Class<T> type) {
     }
 
     public FailoverFeignTarget(Collection<String> servers, int maxBackoffMillis, Class<T> type) {
+        this(servers, maxBackoffMillis, type, RetrySemantics.getSemanticsFor(type));
+    }
+
+    @VisibleForTesting
+    FailoverFeignTarget(
+            Collection<String> servers,
+            int maxBackoffMillis,
+            Class<T> type,
+            RetrySemantics retrySemantics) {
         Preconditions.checkArgument(maxBackoffMillis > 0);
         this.servers = ImmutableList.copyOf(ImmutableSet.copyOf(servers));
         this.type = type;
         this.maxBackoffMillis = maxBackoffMillis;
+        this.retrySemantics = retrySemantics;
     }
 
-    public void sucessfulCall() {
+    public void successfulCall() {
         numSwitches.set(0);
         failuresSinceLastSwitch.set(0);
         startTimeOfFastFailover.set(0);
     }
 
     @Override
     public void continueOrPropagate(RetryableException ex) {
-        boolean isFastFailoverException;
-        if (ex.retryAfter() == null) {
-            // This is the case where we have failed due to networking or other IOException error.
-            isFastFailoverException = false;
-        } else {
-            // This is the case where the server has returned a 503.
-            // This is done when we want to do fast failover because we aren't the leader or we are shutting down.
-            isFastFailoverException = true;
-        }
+        boolean isFastFailoverException = isFastFailoverException(ex);
         synchronized (this) {
+            if (!isFastFailoverException && retrySemantics == RetrySemantics.NEVER_EXCEPT_ON_NON_LEADERS) {
+                throw hardFailOwingToFailureOnLeader(ex);
+            }
```

Comment:
Yep, good spot. It doesn't rely on the failure counts which are presumably why there's synchronization here.